__TESTS__/unit/url/url.test.ts

```diff
@@ -0,0 +1,57 @@
+import {createNewImage} from "../../TestUtils/createCloudinaryImage";
+import {CloudinaryImage} from "../../../src";
+import {Resize} from "../../../src/actions/resize";
+
+
+describe('Tests for URL configuration', () => {
+  it.skip('should accept public_id with special characters', () => {
+    const url = createNewImage('public%id')
+      .toURL();
+    expect(url).toEqual('https://res.cloudinary.com/test123/image/upload/public%25id');
+  });
+  it.skip('should not fail on falsy public_id', () => {
+    const url = createNewImage(void 0)
+      .toURL();
+    expect(url).toEqual(void 0);
+  });
+  it.skip('url should support signature option', () => {
+    const image = new CloudinaryImage('sample', {cloudName:'demo'}, {});
+    const signature = "some-signature";
+    //image.signature(signature)
+    image.resize(Resize.crop().width(100));
+    const url = image.toURL();
+
+    expect(url).toEqual(`https://res.cloudinary.com/demo/image/upload/s--${signature}--/c_crop,w_100/sample`);
+  });
+  describe('window.devicePixelRatio', () => {
+    let options: any;
```

Comment:
All of the devicePixelRatio tests can be deleted,
they are not part of base, and the responsive solution is completely new.